File: ProjectTest/OneLang/Python/OneLang/Test/SelfTestRunner.py
from OneLangStdLib import *
from OneFile import *
import OneLang.Generator.IGenerator as iGen
import OneLang.One.Compiler as comp
import OneLang.Test.PackageStateCapture as packStateCapt
import re

class CompilerHooks:
    def __init__(self, compiler, base_dir):
        self.stage = 0
        self.compiler = compiler
        self.base_dir = base_dir
    
    def after_stage(self, stage_name):
        state = packStateCapt.PackageStateCapture(self.compiler.project_pkg)
        stage_fn = f'''{self.base_dir}/test/artifacts/ProjectTest/OneLang/stages/{self.stage}_{stage_name}.txt'''
        self.stage = self.stage + 1
        stage_summary = state.get_summary()
        expected = OneFile.read_text(stage_fn)
        if stage_summary != expected:
            OneFile.write_text(stage_fn + "_diff.txt", stage_summary)
            raise Error(f'''Stage result differs from expected: {stage_name} -> {stage_fn}''')
        else:
            console.log(f'''[+] Stage passed: {stage_name}''')

class SelfTestRunner:
    def __init__(self, base_dir):
        self.base_dir = base_dir
    
    def run_test(self, generator):
        console.log("[-] SelfTestRunner :: START")
        compiler = comp.Compiler()
        compiler.init(f'''{self.base_dir}packages/''')
        compiler.setup_native_resolver(OneFile.read_text(f'''{self.base_dir}langs/NativeResolvers/typescript.ts'''))
        compiler.new_workspace("OneLang")
        
        proj_dir = f'''{self.base_dir}src/'''
        for file in list(filter(lambda x: x.endswith(".ts"), OneFile.list_files(proj_dir, True))):
            compiler.add_project_file(file, OneFile.read_text(f'''{proj_dir}/{file}'''))
        
        compiler.hooks = CompilerHooks(compiler, self.base_dir)
        
        compiler.process_workspace()
        generated = generator.generate(compiler.project_pkg)
        
        lang_name = generator.get_lang_name()
        ext = f'''.{generator.get_extension()}'''
        
        all_match = True
        for gen_file in generated:
            fn = re.sub("\\.ts$", ext, gen_file.path)
            proj_base = f'''{self.base_dir}test/artifacts/ProjectTest/OneLang'''
            ts_gen_path = f'''{proj_base}/{lang_name}/{fn}'''
            re_gen_path = f'''{proj_base}/{lang_name}_Regen/{fn}'''
            ts_gen_content = OneFile.read_text(ts_gen_path)
            re_gen_content = gen_file.content
            
            if ts_gen_content != re_gen_content:
                OneFile.write_text(re_gen_path, gen_file.content)
                console.error(f'''Content does not match: {gen_file.path}''')
                all_match = False
            else:
                console.log(f'''[+] Content matches: {gen_file.path}''')
        
        console.log("[+} SUCCESS! All generated files are the same" if all_match else "[!] FAIL! Not all files are the same")
        console.log("[-] SelfTestRunner :: DONE")
        return all_match

File: ProjectTest/OneLang/Java/src/main/java/OneLang/Test/SelfTestRunner/SelfTestRunner.java
import java.util.Arrays;
import java.util.regex.Pattern;

public class SelfTestRunner {
    public String baseDir;
    
    public SelfTestRunner(String baseDir)
    {
        this.baseDir = baseDir;
    }
    
    public Boolean runTest(IGenerator generator)
    {
        console.log("[-] SelfTestRunner :: START");
        var compiler = new Compiler();
        compiler.init(this.baseDir + "packages/");
        compiler.setupNativeResolver(OneFile.readText(this.baseDir + "langs/NativeResolvers/typescript.ts"));
        compiler.newWorkspace("OneLang");
        
        var projDir = this.baseDir + "src/";
        for (var file : Arrays.stream(OneFile.listFiles(projDir, true)).filter(x -> x.endsWith(".ts")).toArray(String[]::new))
            compiler.addProjectFile(file, OneFile.readText(projDir + "/" + file));
        
        compiler.hooks = new CompilerHooks(compiler, this.baseDir);
        
        compiler.processWorkspace();
        var generated = generator.generate(compiler.projectPkg);
        
        var langName = generator.getLangName();
        var ext = "." + generator.getExtension();
        
        var allMatch = true;
        for (var genFile : generated) {
            var fn = genFile.path.replaceAll(Pattern.quote("\\.ts$"), ext);
            var projBase = this.baseDir + "test/artifacts/ProjectTest/OneLang";
            var tsGenPath = projBase + "/" + langName + "/" + fn;
            var reGenPath = projBase + "/" + langName + "_Regen/" + fn;
            var tsGenContent = OneFile.readText(tsGenPath);
            var reGenContent = genFile.content;
            
            if (tsGenContent != reGenContent) {
                OneFile.writeText(reGenPath, genFile.content);
                console.error("Content does not match: " + genFile.path);
                allMatch = false;
            }
            else
                console.log("[+] Content matches: " + genFile.path);
        }
        
        console.log(allMatch ? "[+} SUCCESS! All generated files are the same" : "[!] FAIL! Not all files are the same");
        console.log("[-] SelfTestRunner :: DONE");
        return allMatch;
    }
}